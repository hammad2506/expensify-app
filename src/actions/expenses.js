import database from '../firebase/firebase';

export const addExpense = (expense) => (
    {
    type: 'ADD_EXPENSE',
    expense
    }
);

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.user;
        const {description = '', amount = 0, note = '', createdAt = undefined} = expenseData;
        const expense = {description, amount, note, createdAt};
        return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

export const setExpenses = (expenses) => (
    {
        type: 'SET_EXPENSES',
        expenses
    }
);

export const startSetExpenses = () => { 
    return (dispatch, getState) => {
        const uid = getState().auth.user;
        return database.ref(`users/${uid}/expenses`).once('value').then((dataSnapshot) => {
            const expenses = [];
            dataSnapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            }); 
            dispatch(setExpenses(expenses));
        })
    }
}


export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({id}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.user;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense({id}));
        });
    }
}

export const editExpense = ( id, updates= {} ) => ({
    type: 'EDIT_EXPENSE',
    id, 
    updates
});

export const startEditExpense = ( id, updates= {} ) => {
    return (dispatch, getState) => {
        const uid = getState().auth.user;
        return database.ref(`users/${uid}/expenses/${id}`).update({
            ...updates
        }).then(()=>{
            dispatch(editExpense(id, updates));
        });
    }
}
