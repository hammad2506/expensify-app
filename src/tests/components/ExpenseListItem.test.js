import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../dummyData';
import ExpenseListItem from '../../components/ExpenseListItem';

test("Testing Expense List with expense", () => {
    const wrapper =  shallow(<ExpenseListItem {...expenses[1]}/>);
    expect(wrapper).toMatchSnapshot();
});
