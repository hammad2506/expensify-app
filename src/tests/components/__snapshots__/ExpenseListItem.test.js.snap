// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Expense List with expense 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/2"
  >
    <h3>
      Electricity
    </h3>
  </Link>
  <p>
    $
    35.00
     - 
    259200000
  </p>
</div>
`;
